libs/checkpoint/src/store/base.ts

```diff
@@ -0,0 +1,259 @@
+   * @param options Search options.
+   * @param options.filter Key-value pairs to filter results.
+   * @param options.limit Maximum number of items to return (default: 10).
+   * @param options.offset Number of items to skip before returning results (default: 0).
+   * @returns A promise that resolves to a list of items matching the search criteria.
+   */
+  async search(
+    namespacePrefix: string[],
+    options?: {
+      filter?: Record<string, any>;
+      limit?: number;
+      offset?: number;
+    }
+  ): Promise<Item[]> {
+    const optionsWithDefaults = {
+      limit: 10,
+      offset: 0,
+      ...(options || {}),
+    };
+    const batchResults = await this.batch<[SearchOperation]>([
+      { namespacePrefix, ...optionsWithDefaults },
+    ]);
+    return batchResults[0];
+  }
+
+  /**
+   * Store or update an item.
+   * @param namespace Hierarchical path for the item.
+   * @param id Unique identifier within the namespace.
+   * @param value Object containing the item's data.
+   */
+  async put(
+    namespace: string[],
+    id: string,
+    value: Record<string, any>
+  ): Promise<void> {
+    await this.batch<[PutOperation]>([{ namespace, id, value }]);
+  }
+
+  /**
+   * Delete an item.
+   * @param namespace Hierarchical path for the item.
+   * @param id Unique identifier within the namespace.
+   */
+  async delete(namespace: string[], id: string): Promise<void> {
+    await this.batch<[PutOperation]>([{ namespace, id, value: null }]);
+  }
+
+  /**
+   * List and filter namespaces in the store.
+   * @param options Options for listing namespaces.
+   * @param options.prefix Filter namespaces that start with this path.
+   * @param options.suffix Filter namespaces that end with this path.
+   * @param options.maxDepth Return namespaces up to this depth in the hierarchy.
+   * @param options.limit Maximum number of namespaces to return (default 100).
+   * @param options.offset Number of namespaces to skip for pagination (default 0).
+   * @returns A promise that resolves to a list of namespace arrays that match the criteria.
+   */
+  async listNamespaces(options: {
+    prefix?: string[];
+    suffix?: string[];
+    maxDepth?: number;
+    limit?: number;
+    offset?: number;
+  }): Promise<string[][]> {
+    const { prefix, suffix, maxDepth, limit = 100, offset = 0 } = options;
+
+    const matchConditions: MatchCondition[] = [];
+    if (prefix) {
+      matchConditions.push({ matchType: "prefix", path: prefix });
+    }
+    if (suffix) {
+      matchConditions.push({ matchType: "suffix", path: suffix });
+    }
+
+    const op: ListNamespacesOperation = {
+      matchConditions: matchConditions.length > 0 ? matchConditions : undefined,
+      maxDepth,
+      limit,
+      offset,
+    };
+
+    const batchResults = await this.batch<[ListNamespacesOperation]>([op]);
+    return batchResults[0];
+  }
+
+  /**
+   * Stop the store. No-op if not implemented.
+   */
+  stop(): void {
```

Comment:
This is all for batching? What checkpointer uses this?

Reply:
`AsyncBatchedStore` needs it to stop the `while` loop at the end of an execution (if not, it will continue looping until the process exits.